stage1/init/common/kvm_mount.go

```diff
@@ -12,260 +12,89 @@
-	unitName := unit.UnitNamePathEscape(where + ".mount")
-
-	if err := writeUnit(opts, filepath.Join(unitsPath, unitName)); err != nil {
-		return "", err
-	}
-	diag.Printf("mount unit created: %q in %q (what=%q, where=%q)", unitName, unitsPath, what, where)
-
-	return unitName, nil
-}
-
-func writeUnit(opts []*unit.UnitOption, unitPath string) error {
-	unitBytes, err := ioutil.ReadAll(unit.Serialize(opts))
-	if err != nil {
-		return errwrap.Wrap(fmt.Errorf("failed to serialize mount unit file to bytes %q", unitPath), err)
-	}
-
-	err = ioutil.WriteFile(unitPath, unitBytes, 0644)
-	if err != nil {
-		return errwrap.Wrap(fmt.Errorf("failed to create mount unit file %q", unitPath), err)
-	}
-
-	return nil
-}
-
-// PodToSystemdHostMountUnits create host shared remote file system
-// mounts (using e.g. 9p) according to https://www.kernel.org/doc/Documentation/filesystems/9p.txt.
-// Additionally it creates required directories in stage1MntDir and then prepares
-// bind mount unit for each app.
-// "root" parameter is stage1 root filesystem path.
-// appNames are used to create before/required dependency between mount unit and
-// app service units.
-func PodToSystemdHostMountUnits(root string, volumes []types.Volume, appNames []types.ACName, unitsDir string) error {
-	// pod volumes need to mount p9 qemu mount_tags
-	for _, vol := range volumes { // only host shared volumes
-		name := vol.Name.String()
-		mountTag := makeHashFromVolumeName(name)
-
-		// serviceNames for ordering and requirements dependency for apps
-		var serviceNames []string
-		for _, appName := range appNames {
-			serviceNames = append(serviceNames, serviceUnitName(appName))
-		}
-
-		// for host kind we create a mount unit to mount host shared folder
-		if vol.Kind == "host" {
-			// /var/lib/.../pod/run/rootfs/mnt/{mountTag}
-			mountPoint := filepath.Join(root, stage1MntDir, mountTag)
-			err := os.MkdirAll(mountPoint, 0700)
-			if err != nil {
-				return err
-			}
-
-			_, err = installNewMountUnit(root,
-				mountTag, // what (source) in 9p it is a channel tag which equals to volume mountTag
-				filepath.Join(stage1MntDir, name), // where - destination
-				"9p",                            // fsType
-				"trans=virtio",                  // 9p specific options
-				strings.Join(serviceNames, " "), // space separated list of services for unit dependency
-				unitsDir,
-			)
-			if err != nil {
-				return err
-			}
-		}
-	}
-
-	return nil
-}
-
-// AppToSystemdMountUnits prepare bind mount unit for empty or host kind mounting
-// between stage1 rootfs and chrooted filesystem for application
-func AppToSystemdMountUnits(root string, appName types.ACName, volumes []types.Volume, ra *schema.RuntimeApp, unitsDir string) error {
+func MountSharedVolumes(root string, appName types.ACName, volumes []types.Volume, ra *schema.RuntimeApp) error {
 	app := ra.App
-
 	vols := make(map[types.ACName]types.Volume)
 	for _, v := range volumes {
 		vols[v.Name] = v
 	}
```

Comment:
I know it was already like that before, but the mapification of the volume array into a map is done in both calls of `generateMounts` (both in the kvm stage1 and the non-kvm stage1). For factorisation, should `generateMounts` accept an array of volumes as parameter instead?

Both calls of `generateMounts` use the following pattern:

```
	vols := make(map[types.ACName]types.Volume)
	for _, v := range p.Manifest.Volumes {
		vols[v.Name] = v
	}

	mounts := generateMounts(ra, vols)
	for _, m := range mounts {
		vol := vols[m.Volume]
		...
	}
```

Instead, it could be

```
	mounts := generateMounts(ra, vols, func(m Mount, v Volume){
		...
	})
```

(Not necessarily for this PR though)